docs/design-documents/20220408-loading-pipeline-configuration-files.md

```diff
@@ -0,0 +1,270 @@
+# Loading Pipeline Configuration Files
+
+## Summary 
+
+This document describes the design and implementation for loading pipeline 
+configuration files. It details how Conduit loads a configuration file at start.
+
+In order to provide a detailed history of configuration changes, better secret
+handling, and increased confidence in a given configuration, we propose
+adding support for loading and subsequently running pipelines entirely from 
+configuration files loaded into Conduit at its root or a configured directory.
+
+Our solution proposes an immutable Conduit runtime backed by an in-memory DB
+implementation that merges a user-provided configuration file with environment
+variables and CLI flags at runtime to seed Conduit's services with the necessary 
+information for full operation. 
+
+## Goals 
+
+- Fully load and configure pipelines from a file.
+- Properly handle environment variable injection into that file.
+
+### Non Goals
+
+Defining what we don't want out of this design and implementation is equally 
+important as what we _do_ want.
+
+- We don't care about the structure or schema of configuration files.
+- We shouldn't implement key management or encryption for secrets.
+
+## Scope
+
+We are strictly concerned with the loading of pipeline files at start time and 
+how those files are handled through Conduit's full operation cycle.
+
+Configuration file schema and other file-level design choices are out of scope.
+
+This document doesn't make any definitions or statements about how the UI should
+handle these changes. Those should be reserved for later design documents and
+discussion.
+
+## Key Terms 
+
+**pipeline** - a Conduit pipeline with `n` number of Source connectors, 
+`m` number of Processors, and `o` number of Destination connectors.
+
+**production mode** - a mode of operation where Conduit pipelines must be
+statically defined and cannot change after Conduit starts. No UI is served in
+production mode, and only pipeline lifecycle methods can be called over the
+API.
+
+## Context 
+
+Running Conduit in production environments requires a strict set of
+guarantees. Manually configuring a Conduit server in a GUI and keeping those
+high standards places more work on the end user, reduces confidence in the
+pipeline's integrity, and in the worst case could result in corrupted or lost
+data.
+
+This document expands on previous discussions to create a production mode of
+Conduit that either doesn't serve the application GUI or serves a functionally 
+limited version of the GUI instead.
+
+Conduit Issues #32 and #33 describe two different features to address some of 
+these concerns, but there are additional design decisions to make.
+
+## Design 
+
+We propose the creation of a pipeline configuration file that contains all of
+the configuration values needed for `n` number of full pipelines.
```

Comment:
What if we instead have a folder that is scanned on startup and all `.yml`/`.yaml` files in that folder are loaded? Users can still choose to create a single file with [multiple documents](https://yaml.org/spec/1.2.2/#22-structures) by separating them with `---` or split them into separate files.

Reply:
I considered this but after initial feedback on the document I thought that doing less and making it simpler seemed appropriate so I went with just the single file design. I would be open to making this as you described, though.

Reply:
I've updated the design to include language for supporting multiple pipeline configuration files. 👍

Reply:
I've updated the design to include language for supporting multiple pipeline configuration files. 👍

Reply:
The document still largely talks about a single pipeline configuration file (chapter Design, Implementation), if we decided to support multiple configuration files loaded from a single folder then I think we should change the wording.